samples/nrf9160/serial_lte_modem/src/tcpip_proxy/slm_at_udp_proxy.c

```diff
@@ -0,0 +1,608 @@
+ */
+#include <logging/log.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include <string.h>
+#include <net/socket.h>
+#include <modem/modem_info.h>
+#include <net/tls_credentials.h>
+#include "slm_util.h"
+#include "slm_at_host.h"
+#include "slm_at_udp_proxy.h"
+
+LOG_MODULE_REGISTER(udp_proxy, CONFIG_SLM_LOG_LEVEL);
+
+#define THREAD_STACK_SIZE	KB(1)
+#define THREAD_PRIORITY		K_LOWEST_APPLICATION_THREAD_PRIO
+
+/*
+ * Known limitation in this version
+ * - Multiple concurrent
+ * - Receive more than IPv4 MTU one-time
+ * - IPv6 support
+ * - does not support proxy
+ */
+
+/**@brief Proxy operations. */
+enum slm_udp_proxy_operation {
+	AT_SERVER_STOP,
+	AT_CLIENT_DISCONNECT = AT_SERVER_STOP,
+	AT_SERVER_START,
+	AT_CLIENT_CONNECT = AT_SERVER_START
+};
+
+/**@brief Proxy roles. */
+enum slm_udp_proxy_role {
+	AT_UDP_ROLE_CLIENT,
+	AT_UDP_ROLE_SERVER
+};
+
+/**@brief List of supported AT commands. */
+enum slm_udp_proxy_at_cmd_type {
+	AT_UDP_SERVER,
+	AT_UDP_CLIENT,
+	AT_UDP_SEND,
+	AT_UDP_PROXY_MAX
+};
+
+/** forward declaration of cmd handlers **/
+static int handle_at_udp_server(enum at_cmd_type cmd_type);
+static int handle_at_udp_client(enum at_cmd_type cmd_type);
+static int handle_at_udp_send(enum at_cmd_type cmd_type);
+
+/**@brief SLM AT Command list type. */
+static slm_at_cmd_list_t m_udp_proxy_at_list[AT_UDP_PROXY_MAX] = {
+	{AT_UDP_SERVER, "AT#XUDPSVR", handle_at_udp_server},
+	{AT_UDP_CLIENT, "AT#XUDPCLI", handle_at_udp_client},
+	{AT_UDP_SEND, "AT#XUDPSEND", handle_at_udp_send},
+};
+
+static struct k_thread udp_thread;
+static K_THREAD_STACK_DEFINE(udp_thread_stack, THREAD_STACK_SIZE);
+static k_tid_t udp_thread_id;
+
+static struct sockaddr_in remote;
+static int udp_sock;
+
+/* global functions defined in different files */
+void rsp_send(const u8_t *str, size_t len);
+
+/* global variable defined in different files */
+extern struct at_param_list at_param_list;
+extern struct modem_param_info modem_param;
+extern char rsp_buf[CONFIG_AT_CMD_RESPONSE_MAX_LEN];
+
+/** forward declaration of cmd handlers **/
+static int do_udp_server_stop(int error);
+static void udp_thread_func(void *p1, void *p2, void *p3);
+
+static int do_udp_server_start(u16_t port, int sec_tag)
+{
+	int ret = 0;
+	struct sockaddr_in local;
+	int addr_len;
+
+	/* Open socket */
+	if (sec_tag == INVALID_SEC_TAG) {
+		udp_sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+	} else {
+		sprintf(rsp_buf,
+			"#XUDPSVR: DTLS Server not supported\r\n");
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		return -ENOTSUP;
+	}
+	if (udp_sock < 0) {
+		LOG_ERR("socket() failed: %d", -errno);
+		sprintf(rsp_buf, "#XUDPSVR: %d\r\n", -errno);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		return -errno;
+	}
+
+	/* Bind to local port */
+	local.sin_family = AF_INET;
+	local.sin_port = htons(port);
+	ret = modem_info_params_get(&modem_param);
+	if (ret) {
+		LOG_ERR("Unable to obtain modem parameters (%d)", ret);
+		close(udp_sock);
+		return -1;
```

Comment:
Same as for TCP, should return a value from errno space.

Reply:
Updated